### ISSUE TITLE
Elbow curve and other clustering mechanisms for choosing K

### ISSUE THREAD

**Issue by bbengfort** [DistrictDataLabs/yellowbrick#91]
We should implement a multi-model clustering visualization for models that require the specification of a  "k" parameter, in particular, a visualization of the elbow curve, index size, variance thresholds, etc.

**Update**: here are the final things required to wrap up this issue:

- [ ] Add inertia as seen in [sklearn KMeans Equivalent of Elbow Method](https://stackoverflow.com/questions/41540751/sklearn-kmeans-equivalent-of-elbow-method)
- [ ] Parallelize with joblib and n_jobs argument  
- [ ] Find example with clear elbow curve demonstration
- [ ] Finish documentation  
- [x] Add timings to show performance of each k  
- [x] Test on MiniBatchKMeans 
- [x] Add other metrics for both the score and silhouette metric.

**Comment by rebeccabilbro** [comment]
See also: Silhouette Scores

Silhouette analysis can be used to study the separation distance between the resulting clusters. The silhouette plot displays a measure of how close each point in one cluster is to points in the neighboring clusters and thus provides a way to assess parameters like number of clusters visually. 

- http://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_silhouette_analysis.html
- http://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html

**Comment by bbengfort** [comment]
New idea, using the elbow method - create an interaction, that you can click on the graph at the elbow, and set that K to select that model!

**Comment by bbengfort** [comment]
Added a simple `KElbowVisualizer` that uses the silhouette score to draw an elbow curve. This also gave me the opportunity to create the cluster module inside of Yellowbrick. 

![elbow](https://cloud.githubusercontent.com/assets/745966/24279555/efdf83e8-1020-11e7-8ce2-020492b10741.png)

**Comment by bbengfort** [comment]
Left on this one:

- [ ] Parallelize if possible 
- [ ] Find a good example and add to examples notebook 
- [ ] Add more documentation  
- [ ] Add timings to show performance of each k  
- [ ] Test on MiniBatchKMeans 
- [ ] Add other metrics for both the score and silhoutte metric.

**Comment by bbengfort** [comment]
And while we're at it, a Silhouette visualizer:

![sil](https://cloud.githubusercontent.com/assets/745966/24364012/80ff80c4-12df-11e7-94d3-59e99c376ca1.png)

**Comment by bbengfort** [comment]
For silhouette:

- [ ] Do better on the limits of the coefficients 
- [ ] Fix the color selection methodology 
- [ ] Documentation and examples 
- [ ] Find real clustering data set to try it on 
- [ ] Add text for average clustering coefficient score

**Comment by bbengfort** [comment]
Maybe reserve this as a minimum viable visualizer example to get people during PyCon sprints to work on it.

**Comment by bbengfort** [comment]
K-Elbow has been improved with metrics and timings. Close to done, just need to add documentation to this. 

![distortion](https://cloud.githubusercontent.com/assets/745966/26059053/5a11cb1c-394e-11e7-8581-24bdb686802c.png)

![silhouette](https://cloud.githubusercontent.com/assets/745966/26059060/5cdeed2a-394e-11e7-90c2-414442620785.png)

![ch](https://cloud.githubusercontent.com/assets/745966/26059065/5f64b516-394e-11e7-915f-19e43be37573.png)

**Comment by ndanielsen** [comment]
@bbengfort these look great

**Comment by mattharrison** [comment]
Is it possible to add ``inertia_`` as a metric to the elbow plot as seen in 
https://stackoverflow.com/questions/41540751/sklearn-kmeans-equivalent-of-elbow-method

It has a nice feature in that it should be monotonically decreasing and makes interpretation of said plot simple.

**Comment by bbengfort** [comment]
@mattharrison that's a great idea; I'll add that to the wrap up for this issue!

**Comment by AB2017Analytics** [comment]
Hi, I just completed my MS in Data Science and I am new to the industry. Was trying to use KElbowVisualizer to make distortion score elbow for my data, but couldn't make it. I understand make_blobs will create random samples from the given sample number. How can I pass my real samples to plot the same? Please help

**Comment by wagner2010** [comment]
Hi Annnie @AB2017Analytics,
Congratulations on completing your MS degree and welcome to the movement and Yellowbrick.  Don't forget to follow us on Twitter @scikit-yb and to star us here on this repo if you haven't yet.  Thank you for reaching out.  We're sorry to hear of the difficulty you're experiencing.  Would you kindly provide the code you were trying to run and the error you received?  It sounds like you were able to run the sample code in the quick start guide, true?  Are you using pandas to import the data in a Jupyter NB?  We look forward to hearing from you.  You may also find this discussion on loading data to be [helpful](https://www.scikit-yb.org/en/latest/api/datasets/index.html?highlight=csv#api-reference).
Cheers!
Team Yellowbrick

**Comment by AB2017Analytics** [comment]
Hiii, thanks for your response. Yes I loaded my data with pandas using
read_csv. Yes I used the sample code only. But then understood that using
make_blobs will not work with my data. Can you please help me with this?
How can I make the distortion elbow kmeans plot from real dataset? For
example now I used the UCI repository Wisconsin Diagnostic dataset for
breast cancer. Thank you!
